src/server/scripts/Spells/spell_priest.cpp

```diff
@@ -1122,38 +1125,168 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+        spell_pri_prayer_of_mending_aura() : SpellScriptLoader("spell_pri_prayer_of_mending_aura") { }
 
-        class spell_pri_prayer_of_mending_heal_SpellScript : public SpellScript
+        class spell_pri_prayer_of_mending_aura_AuraScript : public AuraScript
         {
-            PrepareSpellScript(spell_pri_prayer_of_mending_heal_SpellScript);
+            PrepareAuraScript(spell_pri_prayer_of_mending_aura_AuraScript);
+
+            bool Validate(SpellInfo const* /*spellInfo*/) override
+            {
+                return ValidateSpellInfo({ SPELL_PRIEST_PRAYER_OF_MENDING_HEAL });
+            }
 
-            void HandleHeal(SpellEffIndex /*effIndex*/)
+            void HandleHeal(ProcEventInfo & /*eventInfo*/)
             {
-                if (Unit* caster = GetOriginalCaster())
+                // Caster: the player(priest) that casters the prayer of mending
+                // Owner: the player that currently has the prayer of mending aura on him
+                Unit* owner = GetUnitOwner();
+                Aura* aura = owner->GetAura(SPELL_PRIEST_PRAYER_OF_MENDING_AURA);
+                if (Unit* caster = GetCaster())
                 {
-                    if (AuraEffect* aurEff = caster->GetAuraEffect(SPELL_PRIEST_T9_HEALING_2P, EFFECT_0))
-                    {
-                        int32 heal = GetHitHeal();
-                        AddPct(heal, aurEff->GetAmount());
-                        SetHitHeal(heal);
-                    }
+                    // Cast the spell to heal the owner
+                    caster->CastSpell(owner, SPELL_PRIEST_PRAYER_OF_MENDING_HEAL);
+
+                    // Only cast jump if stack is higher than 0
+                    if (aura->GetStackAmount() > 1)
+                        owner->CastSpell(owner, SPELL_PRIEST_PRAYER_OF_MENDING_JUMP, true, nullptr, nullptr, caster->GetGUID());
+                    else
+                        aura->Remove();
+                }
+            }
+
+            void Register() override
+            {
+                OnProc += AuraProcFn(spell_pri_prayer_of_mending_aura_AuraScript::HandleHeal);
+            }
+        };
+
+        AuraScript* GetAuraScript() const override
+        {
+            return new spell_pri_prayer_of_mending_aura_AuraScript();
+        }
+};
+
+// 155793 - prayer of mending (Jump) - SPELL_PRIEST_PRAYER_OF_MENDING_JUMP
+class spell_pri_prayer_of_mending_jump : public SpellScriptLoader
+{
+    public:
+        spell_pri_prayer_of_mending_jump() : SpellScriptLoader("spell_pri_prayer_of_mending_jump") { }
+
+        class spell_pri_prayer_of_mending_jump_SpellScript : public spell_pri_prayer_of_mending_SpellScriptBase
+        {
+            PrepareSpellScript(spell_pri_prayer_of_mending_jump_SpellScript);
+
+            void OnTargetSelect(std::list<WorldObject*>& targets)
+            {
+                targets.remove_if([](WorldObject* worldObject)
```

Comment:
This script is incorrect
* (minor) nullptr will be inside the target list
* This is aoe target - you need to pick only one target from the prepopulated list (currently you will cast this on everyone)
* Target is supposed to be a random injured unit (skip full health ones), can hit pets but will always prefer players

Reply:
> (minor) nullptr will be inside the target list
if you mean before, that is why there's a `!worldObject` check inside.
If you mean after, there shouldn't be since `!worldObject` is part of the return (hence if world object is null, it will reutrn true and will be removed)

Reply:
fuck, i wrote that comment wrong. there was supposed to be "nullptr will never be inside the target list"